vm/src/serializer.cpp

```diff
@@ -0,0 +1,33 @@
+#include "serializer.h"
+
+namespace Borges {
+
+Serializer::Serializer(std::vector<double>& b) : buffer_(b)
+{
+}
+
+void
+Serializer::serialize(double x)
+{
+    buffer_.push_back(x);
+}
+
+void
+serialize(Serializer& s, bool p)
+{
+    serialize(s, static_cast<double>(p));
+}
+
+void
+serialize(Serializer& s, double x)
+{
+    s.serialize(x);
+}
+
```

Comment:
still, after our discussions, I don't understand why these cannot or shouldn't be member functions. each function manipulates the state of the Serializer, right? then why whould these be global functions? to align with the deserialize API?

Reply:
No. None of these functions manipulate the state of the serializer directly. Only the two member functions (for serializing doubles or shared_ptrs) can do that. They are thus the only ones that can possible mess up the invariants for the class. All the other functions that use the serializer are implemented *in terms* of these member functions, so they don't need direct access (and hence should not be members);